images/base/Dockerfile

```diff
@@ -117,78 +82,119 @@ RUN echo "Installing Packages ..." \
-    && find /opt/cni/bin -type f -not \( \
-         -iname host-local \
-         -o -iname ptp \
-         -o -iname portmap \
-         -o -iname loopback \
-      \) \
-      -delete
-
-RUN echo "Installing containerd-fuse-overlayfs ..." \
-    && curl -sSL --retry 5 --output /tmp/containerd-fuse-overlayfs.${TARGETARCH}.tgz "${CONTAINERD_FUSE_OVERLAYFS_URL}" \
-    && echo "${CONTAINERD_FUSE_OVERLAYFS_AMD64_SHA256SUM}  /tmp/containerd-fuse-overlayfs.amd64.tgz" | tee /tmp/containerd-fuse-overlayfs.sha256 \
-    && echo "${CONTAINERD_FUSE_OVERLAYFS_ARM64_SHA256SUM}  /tmp/containerd-fuse-overlayfs.arm64.tgz" | tee -a /tmp/containerd-fuse-overlayfs.sha256 \
-    && sha256sum --ignore-missing -c /tmp/containerd-fuse-overlayfs.sha256 \
-    && rm -f /tmp/containerd-fuse-overlayfs.sha256 \
-    && tar -C /usr/local/bin -xzvf /tmp/containerd-fuse-overlayfs.${TARGETARCH}.tgz \
-    && rm -rf /tmp/containerd-fuse-overlayfs.${TARGETARCH}.tgz
+RUN echo "Enabling kubelet and containerd services ... " \
+    && systemctl enable kubelet.service \
+    && systemctl enable containerd.service
 
 RUN echo "Ensuring /etc/kubernetes/manifests" \
     && mkdir -p /etc/kubernetes/manifests
 
 RUN echo "Adjusting systemd-tmpfiles timer" \
     && sed -i /usr/lib/systemd/system/systemd-tmpfiles-clean.timer -e 's#OnBootSec=.*#OnBootSec=1min#'
 
-# squash
+
+# shared stage to setup go version for building binaries
+FROM base as go-build
+COPY --chmod=0755 scripts/third_party/gimme/gimme /usr/local/bin/
+# tools needed at build-time only
+RUN clean-install git make libseccomp-dev gcc libc-dev pkg-config
+# set by makefile to .go-version
+ARG GO_VERSION
+RUN gimme "${GO_VERSION}"
+
+
+# stage for building containerd
+FROM go-build as build-containerd
+ARG GO_VERSION
+ARG CONTAINERD_VERSION="v1.6.21"
+ARG CONTAINERD_CLONE_URL="https://github.com/containerd/containerd"
+# we don't build with optional snapshotters, we never select any of these
+# they're not ideal inside kind anyhow, and we save some disk space
+ARG BUILDTAGS=no_aufs,no_zfs,no_btrfs,no_devmapper
+RUN git clone "${CONTAINERD_CLONE_URL}" /containerd \
+    && cd /containerd \
+    && git checkout "${CONTAINERD_VERSION}" \
+    && eval "$(gimme "${GO_VERSION}")" \
+    && make bin/ctr bin/containerd bin/containerd-shim-runc-v2
+
+# stage for building runc
+FROM go-build as build-runc
+ARG GO_VERSION
+ARG RUNC_VERSION="v1.1.7"
+ARG RUNC_CLONE_URL="https://github.com/opencontainers/runc"
+RUN git clone "${RUNC_CLONE_URL}" /runc \
+    && cd /runc \
+    && git checkout "${RUNC_VERSION}" \
+    && eval "$(gimme "${GO_VERSION}")" \
+    && make runc
+
+# stage for building crictl
+FROM go-build as build-crictl
+ARG GO_VERSION
+ARG CRICTL_CLONE_URL="https://github.com/kubernetes-sigs/cri-tools"
+ARG CRICTL_VERSION="v1.26.1"
+RUN git clone "${CRICTL_CLONE_URL}" /cri-tools \
+    && cd /cri-tools \
+    && git checkout "${CRICTL_VERSION}" \
+    && eval "$(gimme "${GO_VERSION}")" \
```

Comment:
why eval gimme again? is not in the `build` image?

Reply:
We call `gimme ${GO_VERSION}` in the `go-build` stage to fetch the toolchain (which then can be re-used by all the other stages that start `FROM go-build`). But the envs it sets are not persisted.

We call `eval $(gimme "${GO_VERSION}")` in the tooling specific stages before doing builds, in the same shell instance, to ensure that the go toolchain from gimme is used. gimme works by modifying `PATH` etc., when you eval it you're just setting those env to point to the go version it installed locally.

I forgot we could use `source ~/.gimme/envs/go$version.env` instead which might be a tiny bit faster.

https://github.com/travis-ci/gimme/blob/master/README.md